<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!-- NewPage -->
<html lang="en">
<head>
<!-- Generated by javadoc (version 1.7.0_51) on Fri Jun 19 12:54:59 EDT 2015 -->
<title>DoubleVectorIndividual</title>
<meta name="date" content="2015-06-19">
<link rel="stylesheet" type="text/css" href="../../stylesheet.css" title="Style">
</head>
<body>
<script type="text/javascript"><!--
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="DoubleVectorIndividual";
    }
//-->
</script>
<noscript>
<div>JavaScript is disabled on your browser.</div>
</noscript>
<!-- ========= START OF TOP NAVBAR ======= -->
<div class="topNav"><a name="navbar_top">
<!--   -->
</a><a href="#skip-navbar_top" title="Skip navigation links"></a><a name="navbar_top_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../index-all.html">Index</a></li>
<li><a href="../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../ec/vector/ByteVectorIndividual.html" title="class in ec.vector"><span class="strong">Prev Class</span></a></li>
<li><a href="../../ec/vector/FloatVectorIndividual.html" title="class in ec.vector"><span class="strong">Next Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../index.html?ec/vector/DoubleVectorIndividual.html" target="_top">Frames</a></li>
<li><a href="DoubleVectorIndividual.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_top">
<li><a href="../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_top");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field_summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field_detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_top">
<!--   -->
</a></div>
<!-- ========= END OF TOP NAVBAR ========= -->
<!-- ======== START OF CLASS DATA ======== -->
<div class="header">
<div class="subTitle">ec.vector</div>
<h2 title="Class DoubleVectorIndividual" class="title">Class DoubleVectorIndividual</h2>
</div>
<div class="contentContainer">
<ul class="inheritance">
<li>java.lang.Object</li>
<li>
<ul class="inheritance">
<li><a href="../../ec/Individual.html" title="class in ec">ec.Individual</a></li>
<li>
<ul class="inheritance">
<li><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">ec.vector.VectorIndividual</a></li>
<li>
<ul class="inheritance">
<li>ec.vector.DoubleVectorIndividual</li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<div class="description">
<ul class="blockList">
<li class="blockList">
<dl>
<dt>All Implemented Interfaces:</dt>
<dd><a href="../../ec/Prototype.html" title="interface in ec">Prototype</a>, <a href="../../ec/Setup.html" title="interface in ec">Setup</a>, java.io.Serializable, java.lang.Cloneable, java.lang.Comparable</dd>
</dl>
<dl>
<dt>Direct Known Subclasses:</dt>
<dd><a href="../../ec/pso/Particle.html" title="class in ec.pso">Particle</a></dd>
</dl>
<hr>
<br>
<pre>public class <span class="strong">DoubleVectorIndividual</span>
extends <a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></pre>
<div class="block">DoubleVectorIndividual is a VectorIndividual whose genome is an array of
 doubles. Gene values may range from species.mingene(x) to species.maxgene(x),
 inclusive. The default mutation method randomizes genes to new values in this
 range, with <tt>species.mutationProbability</tt>. It can also add gaussian
 noise to the genes, if so directed in the FloatVectorSpecies. If the gaussian
 noise pushes the gene out of range, a new noise value is generated.
 
 <p>
 <P><b>From ec.Individual:</b> 

 <p>In addition to serialization for checkpointing, Individuals may read and write themselves to streams in three ways.

 <ul>
 <li><b>writeIndividual(...,DataOutput)/readIndividual(...,DataInput)</b>&nbsp;&nbsp;&nbsp;This method
 transmits or receives an individual in binary.  It is the most efficient approach to sending
 individuals over networks, etc.  These methods write the evaluated flag and the fitness, then
 call <b>readGenotype/writeGenotype</b>, which you must implement to write those parts of your 
 Individual special to your functions-- the default versions of readGenotype/writeGenotype throw errors.
 You don't need to implement them if you don't plan on using read/writeIndividual.

 <li><b>printIndividual(...,PrintWriter)/readIndividual(...,LineNumberReader)</b>&nbsp;&nbsp;&nbsp;This
 approach transmits or receives an indivdual in text encoded such that the individual is largely readable
 by humans but can be read back in 100% by ECJ as well.  To do this, these methods will encode numbers
 using the <tt>ec.util.Code</tt> class.  These methods are mostly used to write out populations to
 files for inspection, slight modification, then reading back in later on.  <b>readIndividual</b> reads
 in the fitness and the evaluation flag, then calls <b>parseGenotype</b> to read in the remaining individual.
 You are responsible for implementing parseGenotype: the Code class is there to help you.
 <b>printIndividual</b> writes out the fitness and evaluation flag, then calls <b>genotypeToString</b> 
 and printlns the resultant string. You are responsible for implementing the genotypeToString method in such
 a way that parseGenotype can read back in the individual println'd with genotypeToString.  The default form
 of genotypeToString simply calls <b>toString</b>, which you may override instead if you like.  The default
 form of <b>parseGenotype</b> throws an error.  You are not required to implement these methods, but without
 them you will not be able to write individuals to files in a simultaneously computer- and human-readable fashion.

 <li><b>printIndividualForHumans(...,PrintWriter)</b>&nbsp;&nbsp;&nbsp;This
 approach prints an individual in a fashion intended for human consumption only.
 <b>printIndividualForHumans</b> writes out the fitness and evaluation flag, then calls <b>genotypeToStringForHumans</b> 
 and printlns the resultant string. You are responsible for implementing the genotypeToStringForHumans method.
 The default form of genotypeToStringForHumans simply calls <b>toString</b>, which you may override instead if you like
 (though note that genotypeToString's default also calls toString).  You should handle one of these methods properly
 to ensure individuals can be printed by ECJ.
 </ul>

 <p>In general, the various readers and writers do three things: they tell the Fitness to read/write itself,
 they read/write the evaluated flag, and they read/write the gene array.  If you add instance variables to
 a VectorIndividual or subclass, you'll need to read/write those variables as well.
 <b>Default Base</b><br>
 vector.double-vect-ind</div>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../serialized-form.html#ec.vector.DoubleVectorIndividual">Serialized Form</a></dd></dl>
</li>
</ul>
</div>
<div class="summary">
<ul class="blockList">
<li class="blockList">
<!-- =========== FIELD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="field_summary">
<!--   -->
</a>
<h3>Field Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Field Summary table, listing fields, and an explanation">
<caption><span>Fields</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Field and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>double[]</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#genome">genome</a></strong></code>&nbsp;</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>static double</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#MAXIMUM_INTEGER_IN_DOUBLE">MAXIMUM_INTEGER_IN_DOUBLE</a></strong></code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>static java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#P_DOUBLEVECTORINDIVIDUAL">P_DOUBLEVECTORINDIVIDUAL</a></strong></code>&nbsp;</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="fields_inherited_from_class_ec.Individual">
<!--   -->
</a>
<h3>Fields inherited from class&nbsp;ec.<a href="../../ec/Individual.html" title="class in ec">Individual</a></h3>
<code><a href="../../ec/Individual.html#evaluated">evaluated</a>, <a href="../../ec/Individual.html#EVALUATED_PREAMBLE">EVALUATED_PREAMBLE</a>, <a href="../../ec/Individual.html#fitness">fitness</a>, <a href="../../ec/Individual.html#P_INDIVIDUAL">P_INDIVIDUAL</a>, <a href="../../ec/Individual.html#species">species</a></code></li>
</ul>
</li>
</ul>
<!-- ======== CONSTRUCTOR SUMMARY ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_summary">
<!--   -->
</a>
<h3>Constructor Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Constructor Summary table, listing constructors, and an explanation">
<caption><span>Constructors</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colOne" scope="col">Constructor and Description</th>
</tr>
<tr class="altColor">
<td class="colOne"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#DoubleVectorIndividual()">DoubleVectorIndividual</a></strong>()</code>&nbsp;</td>
</tr>
</table>
</li>
</ul>
<!-- ========== METHOD SUMMARY =========== -->
<ul class="blockList">
<li class="blockList"><a name="method_summary">
<!--   -->
</a>
<h3>Method Summary</h3>
<table class="overviewSummary" border="0" cellpadding="3" cellspacing="0" summary="Method Summary table, listing methods, and an explanation">
<caption><span>Methods</span><span class="tabEnd">&nbsp;</span></caption>
<tr>
<th class="colFirst" scope="col">Modifier and Type</th>
<th class="colLast" scope="col">Method and Description</th>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#clamp()">clamp</a></strong>()</code>
<div class="block">Clips each gene value to be within its specified [min,max] range.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.lang.Object</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#clone()">clone</a></strong>()</code>
<div class="block">Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code><a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a></code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#defaultBase()">defaultBase</a></strong>()</code>
<div class="block">Returns the default base for this prototype.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)">defaultCrossover</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                int&nbsp;thread,
                <a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a>&nbsp;ind)</code>
<div class="block">Destructively crosses over the individual with another in some default manner.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#defaultMutate(ec.EvolutionState, int)">defaultMutate</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
             int&nbsp;thread)</code>
<div class="block">Destructively mutates the individual in some default manner.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>double</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#distanceTo(ec.Individual)">distanceTo</a></strong>(<a href="../../ec/Individual.html" title="class in ec">Individual</a>&nbsp;otherInd)</code>
<div class="block">Returns the metric distance to another individual, if such a thing can be measured.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#equals(java.lang.Object)">equals</a></strong>(java.lang.Object&nbsp;ind)</code>
<div class="block">Returns true if I am genetically "equal" to ind.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#genomeLength()">genomeLength</a></strong>()</code>
<div class="block">Returns the length of the gene array.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#genotypeToString()">genotypeToString</a></strong>()</code>
<div class="block">Print to a string the genotype of the Individual in a fashion intended
        to be parsed in again via parseGenotype(...).</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>java.lang.String</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#genotypeToStringForHumans()">genotypeToStringForHumans</a></strong>()</code>
<div class="block">Print to a string the genotype of the Individual in a fashion readable by humans, and not intended
        to be parsed in again.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>java.lang.Object</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#getGenome()">getGenome</a></strong>()</code>
<div class="block">Returns the gene array.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>int</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#hashCode()">hashCode</a></strong>()</code>
<div class="block">Returns a hashcode for the individual, such that individuals which
        are equals(...) each other always return the same
        hash code.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>boolean</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#isInRange()">isInRange</a></strong>()</code>
<div class="block">Returns true if each gene value is within is specified [min,max] range.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#join(java.lang.Object[])">join</a></strong>(java.lang.Object[]&nbsp;pieces)</code>
<div class="block">Joins the n pieces and sets the genome to their concatenation.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>protected void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#parseGenotype(ec.EvolutionState, java.io.LineNumberReader)">parseGenotype</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
             java.io.LineNumberReader&nbsp;reader)</code>
<div class="block">This method is used only by the default version of readIndividual(state,reader),
        and it is intended to be overridden to parse in that part of the individual that
        was outputted in the genotypeToString() method.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#polynomialMutate(ec.EvolutionState, ec.util.MersenneTwisterFast, double, boolean, boolean)">polynomialMutate</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                <a href="../../ec/util/MersenneTwisterFast.html" title="class in ec.util">MersenneTwisterFast</a>&nbsp;random,
                double&nbsp;eta_m,
                boolean&nbsp;alternativePolynomialVersion,
                boolean&nbsp;mutationIsBounded)</code>
<div class="block">This function is broken out to keep it identical to NSGA-II's mutation.c code.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#readGenotype(ec.EvolutionState, java.io.DataInput)">readGenotype</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
            java.io.DataInput&nbsp;dataInput)</code>
<div class="block">Reads in the genotypic information from a DataInput, erasing the previous genotype
        of this Individual.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#reset(ec.EvolutionState, int)">reset</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
     int&nbsp;thread)</code>
<div class="block">Initializes the individual by randomly choosing doubles uniformly from
 mingene to maxgene.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#setGenome(java.lang.Object)">setGenome</a></strong>(java.lang.Object&nbsp;gen)</code>
<div class="block">Sets the gene array.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#setGenomeLength(int)">setGenomeLength</a></strong>(int&nbsp;len)</code>
<div class="block">Sets the genome length.</div>
</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
     <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</code>
<div class="block">This should be used to set up only those things which you share in common
        with all other individuals in your species; individual-specific items
        which make you <i>you</i> should be filled in by Species.newIndividual(...),
        and modified by breeders.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#simulatedBinaryCrossover(ec.util.MersenneTwisterFast, ec.vector.DoubleVectorIndividual, double)">simulatedBinaryCrossover</a></strong>(<a href="../../ec/util/MersenneTwisterFast.html" title="class in ec.util">MersenneTwisterFast</a>&nbsp;random,
                        <a href="../../ec/vector/DoubleVectorIndividual.html" title="class in ec.vector">DoubleVectorIndividual</a>&nbsp;other,
                        double&nbsp;eta_c)</code>&nbsp;</td>
</tr>
<tr class="altColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#split(int[], java.lang.Object[])">split</a></strong>(int[]&nbsp;points,
     java.lang.Object[]&nbsp;pieces)</code>
<div class="block">Splits the genome into n pieces, according to points, which *must* be
 sorted.</div>
</td>
</tr>
<tr class="rowColor">
<td class="colFirst"><code>void</code></td>
<td class="colLast"><code><strong><a href="../../ec/vector/DoubleVectorIndividual.html#writeGenotype(ec.EvolutionState, java.io.DataOutput)">writeGenotype</a></strong>(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
             java.io.DataOutput&nbsp;dataOutput)</code>
<div class="block">Writes the genotypic information to a DataOutput.</div>
</td>
</tr>
</table>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_ec.vector.VectorIndividual">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;ec.vector.<a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></h3>
<code><a href="../../ec/vector/VectorIndividual.html#cloneGenes(java.lang.Object)">cloneGenes</a>, <a href="../../ec/vector/VectorIndividual.html#reset(ec.EvolutionState, int, int)">reset</a>, <a href="../../ec/vector/VectorIndividual.html#size()">size</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_ec.Individual">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;ec.<a href="../../ec/Individual.html" title="class in ec">Individual</a></h3>
<code><a href="../../ec/Individual.html#compareTo(java.lang.Object)">compareTo</a>, <a href="../../ec/Individual.html#merge(ec.EvolutionState, ec.Individual)">merge</a>, <a href="../../ec/Individual.html#printIndividual(ec.EvolutionState, int)">printIndividual</a>, <a href="../../ec/Individual.html#printIndividual(ec.EvolutionState, int, int)">printIndividual</a>, <a href="../../ec/Individual.html#printIndividual(ec.EvolutionState, java.io.PrintWriter)">printIndividual</a>, <a href="../../ec/Individual.html#printIndividualForHumans(ec.EvolutionState, int)">printIndividualForHumans</a>, <a href="../../ec/Individual.html#printIndividualForHumans(ec.EvolutionState, int, int)">printIndividualForHumans</a>, <a href="../../ec/Individual.html#readIndividual(ec.EvolutionState, java.io.DataInput)">readIndividual</a>, <a href="../../ec/Individual.html#readIndividual(ec.EvolutionState, java.io.LineNumberReader)">readIndividual</a>, <a href="../../ec/Individual.html#toString()">toString</a>, <a href="../../ec/Individual.html#writeIndividual(ec.EvolutionState, java.io.DataOutput)">writeIndividual</a></code></li>
</ul>
<ul class="blockList">
<li class="blockList"><a name="methods_inherited_from_class_java.lang.Object">
<!--   -->
</a>
<h3>Methods inherited from class&nbsp;java.lang.Object</h3>
<code>finalize, getClass, notify, notifyAll, wait, wait, wait</code></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
<div class="details">
<ul class="blockList">
<li class="blockList">
<!-- ============ FIELD DETAIL =========== -->
<ul class="blockList">
<li class="blockList"><a name="field_detail">
<!--   -->
</a>
<h3>Field Detail</h3>
<a name="P_DOUBLEVECTORINDIVIDUAL">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>P_DOUBLEVECTORINDIVIDUAL</h4>
<pre>public static final&nbsp;java.lang.String P_DOUBLEVECTORINDIVIDUAL</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.DoubleVectorIndividual.P_DOUBLEVECTORINDIVIDUAL">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="MAXIMUM_INTEGER_IN_DOUBLE">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>MAXIMUM_INTEGER_IN_DOUBLE</h4>
<pre>public static final&nbsp;double MAXIMUM_INTEGER_IN_DOUBLE</pre>
<dl><dt><span class="strong">See Also:</span></dt><dd><a href="../../constant-values.html#ec.vector.DoubleVectorIndividual.MAXIMUM_INTEGER_IN_DOUBLE">Constant Field Values</a></dd></dl>
</li>
</ul>
<a name="genome">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>genome</h4>
<pre>public&nbsp;double[] genome</pre>
</li>
</ul>
</li>
</ul>
<!-- ========= CONSTRUCTOR DETAIL ======== -->
<ul class="blockList">
<li class="blockList"><a name="constructor_detail">
<!--   -->
</a>
<h3>Constructor Detail</h3>
<a name="DoubleVectorIndividual()">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>DoubleVectorIndividual</h4>
<pre>public&nbsp;DoubleVectorIndividual()</pre>
</li>
</ul>
</li>
</ul>
<!-- ============ METHOD DETAIL ========== -->
<ul class="blockList">
<li class="blockList"><a name="method_detail">
<!--   -->
</a>
<h3>Method Detail</h3>
<a name="defaultBase()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>defaultBase</h4>
<pre>public&nbsp;<a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;defaultBase()</pre>
<div class="block"><strong>Description copied from interface:&nbsp;<code><a href="../../ec/Prototype.html#defaultBase()">Prototype</a></code></strong></div>
<div class="block">Returns the default base for this prototype.
        This should generally be implemented by building off of the static base()
        method on the DefaultsForm object for the prototype's package. This should
        be callable during setup(...).</div>
</li>
</ul>
<a name="clone()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>clone</h4>
<pre>public&nbsp;java.lang.Object&nbsp;clone()</pre>
<div class="block"><strong>Description copied from interface:&nbsp;<code><a href="../../ec/Prototype.html#clone()">Prototype</a></code></strong></div>
<div class="block">Creates a new individual cloned from a prototype,
        and suitable to begin use in its own evolutionary
        context.

        <p>Typically this should be a full "deep" clone.
        However, you may share certain elements with other objects
        rather than clone hem, depending on the situation:

        <p>
        <ul>
        <li>If you hold objects which are shared with other instances,
        don't clone them.
        <li>If you hold objects which must be unique, clone them.
        <li>If you hold objects which were given to you as a gesture
        of kindness, and aren't owned by you, you probably shouldn't clone
        them.
        <li> DON'T attempt to clone: Singletons, Cliques, or Groups.
        <li>Arrays are not cloned automatically; you may need to
        clone an array if you're not sharing it with other instances.
        Arrays have the nice feature of being copyable by calling clone()
        on them.
        </ul>

        <p><b>Implementations.</b>

        <ul>
        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are abstract, then you should not declare clone().

        <li>If no ancestor of yours implements clone(),
        and you have no need to do clone deeply,
        and you are <b>not</b> abstract, then you should implement
        it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         return super.clone();
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     }
        </pre></tt>

        <li>If no ancestor of yours implements clone(), but you
        need to deep-clone some things, then you should implement it
        as follows:

        <p>
        <tt><pre>
 public Object clone() 
     {
     try
         { 
         MyObject myobj = (MyObject) (super.clone());

         // put your deep-cloning code here...
         }
     catch ((CloneNotSupportedException e)
         { throw new InternalError(); } // never happens
     return myobj;
     } 
        </pre></tt>

        <li>If an ancestor has implemented clone(), and you also need
        to deep clone some things, then you should implement it as follows:

        <p>
        <tt><pre>
 public Object clone() 
     { 
     MyObject myobj = (MyObject) (super.clone());

     // put your deep-cloning code here...

     return myobj;
     } 
        </pre></tt>
        </ul></div>
<dl>
<dt><strong>Specified by:</strong></dt>
<dd><code><a href="../../ec/Prototype.html#clone()">clone</a></code>&nbsp;in interface&nbsp;<code><a href="../../ec/Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Individual.html#clone()">clone</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
</dl>
</li>
</ul>
<a name="setup(ec.EvolutionState, ec.util.Parameter)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setup</h4>
<pre>public&nbsp;void&nbsp;setup(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
         <a href="../../ec/util/Parameter.html" title="class in ec.util">Parameter</a>&nbsp;base)</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#setup(ec.EvolutionState, ec.util.Parameter)">Individual</a></code></strong></div>
<div class="block">This should be used to set up only those things which you share in common
        with all other individuals in your species; individual-specific items
        which make you <i>you</i> should be filled in by Species.newIndividual(...),
        and modified by breeders.</div>
<dl>
<dt><strong>Specified by:</strong></dt>
<dd><code><a href="../../ec/Prototype.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../ec/Prototype.html" title="interface in ec">Prototype</a></code></dd>
<dt><strong>Specified by:</strong></dt>
<dd><code><a href="../../ec/Setup.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</a></code>&nbsp;in interface&nbsp;<code><a href="../../ec/Setup.html" title="interface in ec">Setup</a></code></dd>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Individual.html#setup(ec.EvolutionState, ec.util.Parameter)">setup</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
<dt><span class="strong">See Also:</span></dt><dd><a href="../../ec/Prototype.html#setup(ec.EvolutionState, ec.util.Parameter)"><code>Prototype.setup(EvolutionState,Parameter)</code></a></dd></dl>
</li>
</ul>
<a name="defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>defaultCrossover</h4>
<pre>public&nbsp;void&nbsp;defaultCrossover(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                    int&nbsp;thread,
                    <a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a>&nbsp;ind)</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/vector/VectorIndividual.html#defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)">VectorIndividual</a></code></strong></div>
<div class="block">Destructively crosses over the individual with another in some default manner.  In most
        implementations provided in ECJ, one-, two-, and any-point crossover is done with a 
        for loop, rather than a possibly more efficient approach like arrayCopy().  The disadvantage
        is that arrayCopy() takes advantage of a CPU's bulk copying.  The advantage is that arrayCopy()
        would require a scratch array, so you'd be allocing and GCing an array for every crossover.
        Dunno which is more efficient.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#defaultCrossover(ec.EvolutionState, int, ec.vector.VectorIndividual)">defaultCrossover</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="split(int[], java.lang.Object[])">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>split</h4>
<pre>public&nbsp;void&nbsp;split(int[]&nbsp;points,
         java.lang.Object[]&nbsp;pieces)</pre>
<div class="block">Splits the genome into n pieces, according to points, which *must* be
 sorted. pieces.length must be 1 + points.length</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#split(int[], java.lang.Object[])">split</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="join(java.lang.Object[])">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>join</h4>
<pre>public&nbsp;void&nbsp;join(java.lang.Object[]&nbsp;pieces)</pre>
<div class="block">Joins the n pieces and sets the genome to their concatenation.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#join(java.lang.Object[])">join</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="defaultMutate(ec.EvolutionState, int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>defaultMutate</h4>
<pre>public&nbsp;void&nbsp;defaultMutate(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                 int&nbsp;thread)</pre>
<div class="block">Destructively mutates the individual in some default manner. The default
 form simply randomizes genes to a uniform distribution from the min and
 max of the gene values. It can also add gaussian noise to the genes, if
 so directed in the FloatVectorSpecies. If the gaussian noise pushes the
 gene out of range, a new noise value is generated.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#defaultMutate(ec.EvolutionState, int)">defaultMutate</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="polynomialMutate(ec.EvolutionState, ec.util.MersenneTwisterFast, double, boolean, boolean)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>polynomialMutate</h4>
<pre>public&nbsp;void&nbsp;polynomialMutate(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                    <a href="../../ec/util/MersenneTwisterFast.html" title="class in ec.util">MersenneTwisterFast</a>&nbsp;random,
                    double&nbsp;eta_m,
                    boolean&nbsp;alternativePolynomialVersion,
                    boolean&nbsp;mutationIsBounded)</pre>
<div class="block">This function is broken out to keep it identical to NSGA-II's mutation.c code. eta_m is the distribution
        index.</div>
</li>
</ul>
<a name="simulatedBinaryCrossover(ec.util.MersenneTwisterFast, ec.vector.DoubleVectorIndividual, double)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>simulatedBinaryCrossover</h4>
<pre>public&nbsp;void&nbsp;simulatedBinaryCrossover(<a href="../../ec/util/MersenneTwisterFast.html" title="class in ec.util">MersenneTwisterFast</a>&nbsp;random,
                            <a href="../../ec/vector/DoubleVectorIndividual.html" title="class in ec.vector">DoubleVectorIndividual</a>&nbsp;other,
                            double&nbsp;eta_c)</pre>
</li>
</ul>
<a name="reset(ec.EvolutionState, int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>reset</h4>
<pre>public&nbsp;void&nbsp;reset(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
         int&nbsp;thread)</pre>
<div class="block">Initializes the individual by randomly choosing doubles uniformly from
 mingene to maxgene.</div>
<dl>
<dt><strong>Specified by:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#reset(ec.EvolutionState, int)">reset</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="hashCode()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>hashCode</h4>
<pre>public&nbsp;int&nbsp;hashCode()</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#hashCode()">Individual</a></code></strong></div>
<div class="block">Returns a hashcode for the individual, such that individuals which
        are equals(...) each other always return the same
        hash code.</div>
<dl>
<dt><strong>Specified by:</strong></dt>
<dd><code><a href="../../ec/Individual.html#hashCode()">hashCode</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
</dl>
</li>
</ul>
<a name="genotypeToStringForHumans()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genotypeToStringForHumans</h4>
<pre>public&nbsp;java.lang.String&nbsp;genotypeToStringForHumans()</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#genotypeToStringForHumans()">Individual</a></code></strong></div>
<div class="block">Print to a string the genotype of the Individual in a fashion readable by humans, and not intended
        to be parsed in again.  The fitness and evaluated flag should not be included.  The default form
        simply calls toString(), but you'll probably want to override this to something else.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Individual.html#genotypeToStringForHumans()">genotypeToStringForHumans</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
</dl>
</li>
</ul>
<a name="genotypeToString()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genotypeToString</h4>
<pre>public&nbsp;java.lang.String&nbsp;genotypeToString()</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#genotypeToString()">Individual</a></code></strong></div>
<div class="block">Print to a string the genotype of the Individual in a fashion intended
        to be parsed in again via parseGenotype(...).
        The fitness and evaluated flag should not be included.  The default form
        simply calls toString(), which is almost certainly wrong, and you'll probably want to override
        this to something else.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Individual.html#genotypeToString()">genotypeToString</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
</dl>
</li>
</ul>
<a name="parseGenotype(ec.EvolutionState, java.io.LineNumberReader)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>parseGenotype</h4>
<pre>protected&nbsp;void&nbsp;parseGenotype(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                 java.io.LineNumberReader&nbsp;reader)
                      throws java.io.IOException</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#parseGenotype(ec.EvolutionState, java.io.LineNumberReader)">Individual</a></code></strong></div>
<div class="block">This method is used only by the default version of readIndividual(state,reader),
        and it is intended to be overridden to parse in that part of the individual that
        was outputted in the genotypeToString() method.  The default version of this method
        exits the program with an "unimplemented" error.  You'll want to override this method,
        or to override readIndividual(...) to not use this method.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Individual.html#parseGenotype(ec.EvolutionState, java.io.LineNumberReader)">parseGenotype</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="equals(java.lang.Object)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>equals</h4>
<pre>public&nbsp;boolean&nbsp;equals(java.lang.Object&nbsp;ind)</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#equals(java.lang.Object)">Individual</a></code></strong></div>
<div class="block">Returns true if I am genetically "equal" to ind.  This should
        mostly be interpreted as saying that we are of the same class
        and that we hold the same data. It should NOT be a pointer comparison.</div>
<dl>
<dt><strong>Specified by:</strong></dt>
<dd><code><a href="../../ec/Individual.html#equals(java.lang.Object)">equals</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
</dl>
</li>
</ul>
<a name="getGenome()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>getGenome</h4>
<pre>public&nbsp;java.lang.Object&nbsp;getGenome()</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/vector/VectorIndividual.html#getGenome()">VectorIndividual</a></code></strong></div>
<div class="block">Returns the gene array.  If you know the type of the array, you can cast it and work on
        it directly.  Otherwise, you can still manipulate it in general, because arrays (like
        all objects) respond to clone() and can be manipulated with arrayCopy without bothering
        with their type.  This might be useful in creating special generalized crossover operators
        -- we apologize in advance for the fact that Java doesn't have a template system.  :-( 
        The default version returns null.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#getGenome()">getGenome</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="setGenome(java.lang.Object)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setGenome</h4>
<pre>public&nbsp;void&nbsp;setGenome(java.lang.Object&nbsp;gen)</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/vector/VectorIndividual.html#setGenome(java.lang.Object)">VectorIndividual</a></code></strong></div>
<div class="block">Sets the gene array.  See getGenome().  The default version does nothing.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#setGenome(java.lang.Object)">setGenome</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="genomeLength()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>genomeLength</h4>
<pre>public&nbsp;int&nbsp;genomeLength()</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/vector/VectorIndividual.html#genomeLength()">VectorIndividual</a></code></strong></div>
<div class="block">Returns the length of the gene array.  By default, this method returns 0.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#genomeLength()">genomeLength</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="writeGenotype(ec.EvolutionState, java.io.DataOutput)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>writeGenotype</h4>
<pre>public&nbsp;void&nbsp;writeGenotype(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                 java.io.DataOutput&nbsp;dataOutput)
                   throws java.io.IOException</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#writeGenotype(ec.EvolutionState, java.io.DataOutput)">Individual</a></code></strong></div>
<div class="block">Writes the genotypic information to a DataOutput.  Largely called by writeIndividual(), and
        nothing else.  The default simply throws an error.  Various subclasses of Individual override this as
        appropriate. For example, if your custom individual's genotype consists of an array of 
        integers, you might do this:

 <pre><tt>
 dataOutput.writeInt(integers.length);
 for(int x=0;x<integers.length;x++)
     dataOutput.writeInt(integers[x]);
 </tt></pre></div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Individual.html#writeGenotype(ec.EvolutionState, java.io.DataOutput)">writeGenotype</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="readGenotype(ec.EvolutionState, java.io.DataInput)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>readGenotype</h4>
<pre>public&nbsp;void&nbsp;readGenotype(<a href="../../ec/EvolutionState.html" title="class in ec">EvolutionState</a>&nbsp;state,
                java.io.DataInput&nbsp;dataInput)
                  throws java.io.IOException</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#readGenotype(ec.EvolutionState, java.io.DataInput)">Individual</a></code></strong></div>
<div class="block">Reads in the genotypic information from a DataInput, erasing the previous genotype
        of this Individual.  Largely called by readIndividual(), and nothing else.  
        If you are trying to <i>create</i> an Individual
        from information read in from a stream or DataInput,
        see the various newIndividual() methods in Species.
        The default simply throws an error.  Various subclasses of Individual override this as
        appropriate.  For example, if your custom individual's genotype consists of an array of 
        integers, you might do this:

 <pre><tt>
 integers = new int[dataInput.readInt()];
 for(int x=0;x<integers.length;x++)
     integers[x] = dataInput.readInt();
 </tt></pre></div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Individual.html#readGenotype(ec.EvolutionState, java.io.DataInput)">readGenotype</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
<dt><span class="strong">Throws:</span></dt>
<dd><code>java.io.IOException</code></dd></dl>
</li>
</ul>
<a name="clamp()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>clamp</h4>
<pre>public&nbsp;void&nbsp;clamp()</pre>
<div class="block">Clips each gene value to be within its specified [min,max] range.  
        NaN is presently considered in range but the behavior of this method
        should be assumed to be unspecified on encountering NaN.</div>
</li>
</ul>
<a name="setGenomeLength(int)">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>setGenomeLength</h4>
<pre>public&nbsp;void&nbsp;setGenomeLength(int&nbsp;len)</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/vector/VectorIndividual.html#setGenomeLength(int)">VectorIndividual</a></code></strong></div>
<div class="block">Sets the genome length.  If the length is longer, then it is filled with a default value (likely 0 or false).
        This may or may not be a valid value -- you will need to set appropriate values here. 
        The default implementation does nothing; but all subclasses in ECJ implement a subset of this.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/vector/VectorIndividual.html#setGenomeLength(int)">setGenomeLength</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/vector/VectorIndividual.html" title="class in ec.vector">VectorIndividual</a></code></dd>
</dl>
</li>
</ul>
<a name="isInRange()">
<!--   -->
</a>
<ul class="blockList">
<li class="blockList">
<h4>isInRange</h4>
<pre>public&nbsp;boolean&nbsp;isInRange()</pre>
<div class="block">Returns true if each gene value is within is specified [min,max] range.
        NaN is presently considered in range but the behavior of this method
        should be assumed to be unspecified on encountering NaN.</div>
</li>
</ul>
<a name="distanceTo(ec.Individual)">
<!--   -->
</a>
<ul class="blockListLast">
<li class="blockList">
<h4>distanceTo</h4>
<pre>public&nbsp;double&nbsp;distanceTo(<a href="../../ec/Individual.html" title="class in ec">Individual</a>&nbsp;otherInd)</pre>
<div class="block"><strong>Description copied from class:&nbsp;<code><a href="../../ec/Individual.html#distanceTo(ec.Individual)">Individual</a></code></strong></div>
<div class="block">Returns the metric distance to another individual, if such a thing can be measured.
        Subclassess of Individual should implement this if it exists for their representation.
        The default implementation here, which isn't very helpful, returns 0 if the individuals are equal
        and infinity if they are not.</div>
<dl>
<dt><strong>Overrides:</strong></dt>
<dd><code><a href="../../ec/Individual.html#distanceTo(ec.Individual)">distanceTo</a></code>&nbsp;in class&nbsp;<code><a href="../../ec/Individual.html" title="class in ec">Individual</a></code></dd>
</dl>
</li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
</div>
<!-- ========= END OF CLASS DATA ========= -->
<!-- ======= START OF BOTTOM NAVBAR ====== -->
<div class="bottomNav"><a name="navbar_bottom">
<!--   -->
</a><a href="#skip-navbar_bottom" title="Skip navigation links"></a><a name="navbar_bottom_firstrow">
<!--   -->
</a>
<ul class="navList" title="Navigation">
<li><a href="../../overview-summary.html">Overview</a></li>
<li><a href="package-summary.html">Package</a></li>
<li class="navBarCell1Rev">Class</li>
<li><a href="package-tree.html">Tree</a></li>
<li><a href="../../deprecated-list.html">Deprecated</a></li>
<li><a href="../../index-all.html">Index</a></li>
<li><a href="../../help-doc.html">Help</a></li>
</ul>
</div>
<div class="subNav">
<ul class="navList">
<li><a href="../../ec/vector/ByteVectorIndividual.html" title="class in ec.vector"><span class="strong">Prev Class</span></a></li>
<li><a href="../../ec/vector/FloatVectorIndividual.html" title="class in ec.vector"><span class="strong">Next Class</span></a></li>
</ul>
<ul class="navList">
<li><a href="../../index.html?ec/vector/DoubleVectorIndividual.html" target="_top">Frames</a></li>
<li><a href="DoubleVectorIndividual.html" target="_top">No Frames</a></li>
</ul>
<ul class="navList" id="allclasses_navbar_bottom">
<li><a href="../../allclasses-noframe.html">All Classes</a></li>
</ul>
<div>
<script type="text/javascript"><!--
  allClassesLink = document.getElementById("allclasses_navbar_bottom");
  if(window==top) {
    allClassesLink.style.display = "block";
  }
  else {
    allClassesLink.style.display = "none";
  }
  //-->
</script>
</div>
<div>
<ul class="subNavList">
<li>Summary:&nbsp;</li>
<li>Nested&nbsp;|&nbsp;</li>
<li><a href="#field_summary">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_summary">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_summary">Method</a></li>
</ul>
<ul class="subNavList">
<li>Detail:&nbsp;</li>
<li><a href="#field_detail">Field</a>&nbsp;|&nbsp;</li>
<li><a href="#constructor_detail">Constr</a>&nbsp;|&nbsp;</li>
<li><a href="#method_detail">Method</a></li>
</ul>
</div>
<a name="skip-navbar_bottom">
<!--   -->
</a></div>
<!-- ======== END OF BOTTOM NAVBAR ======= -->
</body>
</html>
